### PR TITLE
Improve test performance

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
can you mention the perf benefit we get for certain tests?

File: src/Compilers/Core/Portable/InternalUtilities/Hash.cs [hunk 1]
```diff
@@ -87,7 +87,7 @@ internal static int CombineValues<T>(T[]? values, int maxItemsToHash = int.MaxVa
                 // Should end up with a constrained virtual call to object.GetHashCode (i.e. avoid boxing where possible).
                 if (value != null)
                 {
-                    hashCode = Hash.Combine(value.GetHashCode(), hashCode);
+                    hashCode = Hash.Combine(EqualityComparer<T>.Default.GetHashCode(value), hashCode);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/InternalUtilities/Hash.cs:90]: yikes.  i thought this issue was fixed by the runtime a long time ago :(

File: src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs [hunk 2]
```diff
@@ -193,6 +193,11 @@ public void Serialize(ObjectWriter writer, CancellationToken cancellationToken)
                     continue;
                 }
 
+                if (Equals(value, optionKey.Option.DefaultValue))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs:196]: nice.  possibly worth a comment.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs [hunk 3]
```diff
@@ -63,6 +63,32 @@ internal partial struct NamingStyle
...
 
+        public override bool Equals(object obj)
+        {
+            return obj is NamingStyle other
+                && Equals(other);
+        }
+
+        public bool Equals(NamingStyle other)
+        {
+            return ID == other.ID
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs:74]: ..

what's the point of a guid here :)   if we have a guid, isn't htat enough for equality ;-)

(note: i'm find with keeping the pr like this.  but if we have naming styles with the same guids and different values, i'm going to be a sad panda).

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs [hunk 4]
```diff
@@ -63,6 +63,32 @@ internal partial struct NamingStyle
...
+        }
+
+        public override int GetHashCode()
+        {
+            return Hash.Combine(ID.GetHashCode(),
+                Hash.Combine(Name?.GetHashCode() ?? 0,
+                    Hash.Combine(Prefix?.GetHashCode() ?? 0,
+                        Hash.Combine(Suffix?.GetHashCode() ?? 0,
+                            Hash.Combine(WordSeparator?.GetHashCode() ?? 0,
+                                (int)CapitalizationScheme)))));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs:89]: is HashCode.Combine not available at this layer?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs [hunk 5]
```diff
@@ -437,7 +462,7 @@ public override bool Equals(object obj)
                 => obj is ModifierKind kind && Equals(kind);
 
             public override int GetHashCode()
-                => ModifierKindWrapper.GetHashCode();
+                => (int)ModifierKindWrapper;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:465]: ughhhhhhhhh

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs [hunk 6]
```diff
@@ -182,6 +182,32 @@ private bool AllMatches<TSymbolMatcher>(ImmutableArray<TSymbolMatcher> matchers,
...
+        }
+
+        public override int GetHashCode()
+        {
+            return Hash.Combine(ID.GetHashCode(),
+                Hash.Combine(Name.GetHashCode(),
+                    Hash.Combine(Hash.CombineValues(ApplicableSymbolKindList),
+                        Hash.Combine(Hash.CombineValues(ApplicableAccessibilityList),
+                            Hash.CombineValues(RequiredModifierList)))));
+        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:209]: so why does this help?  clearly we didn't have equals/gethashcode here before... so we coudlnt' have been depending on htat behavior.  so how do this help out?

File: src/Compilers/Core/Portable/InternalUtilities/Hash.cs (hunk 1 above)
Comment (cston) [inline src/Compilers/Core/Portable/InternalUtilities/Hash.cs:90]: >EqualityComparer<T>.Default.GetHashCode(value) [](start = 44, length = 46)

What is the reason for this change?

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/InternalUtilities/Hash.cs:90]: > What is the reason for this change?

Yes, please, add a comment explaining the reason so that the change won't be undone unintentionally.

---
In reply to: [403348616](https://github.com/dotnet/roslyn/pull/43060#discussion_r403348616) [](ancestors = 403348616)

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs:89]: HashCode.Combine is not in any layer IIRC (netcoreapp only)

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs (hunk 6 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:209]: Way up above we serialized the whole tree to XElement and compared those.

File: src/Compilers/Core/Portable/InternalUtilities/Hash.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/Hash.cs:90]: I'll revert this change. PerfView is showing allocations of NamingStyle from this method even though the new NamingStyle.GetHashCode implementation is called. I'm not sure why, but the savings from this specific change are not great enough to warrant deep investigation.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs:74]: Not sure why, but this matches what we did before and the perf issue is gone.

File: src/Compilers/Core/Portable/InternalUtilities/Hash.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/Hash.cs:90]: ➡️ Discussion for this item is in https://github.com/dotnet/roslyn/pull/43060#discussion_r403348616

**Comment by sharwell** [comment]
> can you mention the perf benefit we get for certain tests?

Not particularly, I didn't measure the changes in isolation, plus the characteristics on the build server are very different from the ones locally.

File: src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs:196]: @mavasani @jasonmalinowski Is this a valid optimization? If so, it's one of the bigger wins in this pull request.

Comment (mavasani) [reply src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs:196]: Looks reasonable to me - from quick glance at the uses of `_serializableOptionValues`, it doesn't seem checksum validation will be affected.

File: src/Compilers/Core/Portable/InternalUtilities/Hash.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/Hash.cs]: So the claim about a constrained `object.GetHashCode()` is somewhat correct. There are some scenarios where it might not work (e.g. tiered compilation), but the big case is it fails to account for this:

https://github.com/dotnet/runtime/blob/c21c7fd280aed2ba12ed3807c1082e8aba95a602/src/coreclr/src/System.Private.CoreLib/src/System/Collections/Generic/ComparerHelpers.cs#L78-L108

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:209]: Ah, yikes.  thanks.

File: src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs:196]: This one is causing test failures. I'm going to back it out so we can get the rest in.
